Command class readme

jonahx opened the pull request:
### What does this PR do?

Adds README for `app/domain` and `CommandClass`

jonahx commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+
+## Introduction
+
+This document explains the coding standards used in `app/domain`, and the
```
I definitely could have put this under `lib`.  That may have been the better choice.  I don't have a strong feeling one way or another and would have no objection to moving it.

micahlee commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+    - By contrast, a value object would use a name like `github.issues`.  The
+      "fetch" indicates the network side effect and possibility of error.
+
+Other principles include:
```
This would benefit from examples, both for good and bad cases.

The authentication audit events would be a good example for avoiding inheritance, also the equality operator.

The use of database lookups (though subtle with sequel) are common "bad" examples for immutability.

Not sure about naming examples off-hand, but this might benefit the most where a few examples communicate the "spirit" of the principal more clearly than the description itself.

micahlee commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+
+- Hardcoding dependencies that should be injected.
+  - For example: Directly referencing `Rails.logger` or a model class.
+  - All code with side effects should be injected as a depenency.
```
```suggestion
  - All code with side effects should be injected as a dependency.
```

micahlee commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+
+So we now have two options for vanilla ruby alternatives to the `CommandClass`:
+
+1. One which is coneptually simple but cluttered by passing runtime arguments
```
```suggestion
1. One which is conceptually simple but cluttered by passing runtime arguments
```

micahlee commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+1. One which avoids that clutter but at the expense of other complexity and
+   boilerplate.
+
+The `CommandClass` is simply a third option, which takes the second approach --
```
Would it be difficult to include an example of code that you would _not_ convert to a command class?

jonahx commented:
@micahlee Only just now saw your comments.  Really appreciate them.  I'll try to think of examples per your suggestions.

orenbm commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+- **Command objects**
+  - These go by many names: service objects, interactors, use cases
+  - Coordinates high-level logic involving multiple collaborators
+    - For example: `CreateUser` might need to coordinate user input validation,
+      a database transaction, a log entry, and a confirmation email
```
i would add here that `CreateUser` will consume some of its logic from other Command objects with smaller resplonsibility

orenbm commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+
+You'll encounter two gems when working in Conjur:
+
+- `command_class` for writing command objects
```
do you want to move this repo to a cyberark one?

jonahx replied:
Well, all my work on it was done on my own time.  If there's a pressing need to I suppose I would, but it's really no different from our perspective than using any 3rd party gem and we depend on hundreds of them.

orenbm commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+
+- `command_class` for writing command objects
+  - This is still used extensively.  Read below for more information.
+- `dry-struct` for value objects
+  - While this is a good gem, I've stopped using inside Conjur for the reasons
+    listed above
+  - I now write value objects using plain ruby
```
do we have an effort to remove this altogether? how hard do you think it will be?

jonahx replied:
Probably not that hard.  I don't think there are that many instances but I'd have to check.

orenbm commented on app/domain/README.md:
```diff
@@ -0,0 +1,296 @@
+- Avoid putting too many responsibilities into a command class.  Its single
+  verb phrase name should naturally describe it.
+  - Split up a `CommandClass` when needed
+- Avoid putting too little in one.  Not everything should be a `CommandClass`.
+  Keep in mind why we write them.
```
i'd like to get an example here as this is something that i never really understood. From my point of view, if you have a responsibility that needs to be done in several places then it makes sense to make it a command class. i have an example in another PR which i will link to once it is merged.

orenbm replied:
here is the example. we have this class: https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/extract_container_name.rb

it is used in both [here](https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/inject_client_cert.rb#L177) and [here](https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/validate_pod_request.rb#L88)

you will see that the command class has only a dependency on `Rails.logger` which may point that it should not be a command class. However, it does have some logic and constants that i think should not be implemented in both places. 

i'd like to get your thoughts on this, and even more important - extract a rule of thumb for this README.

jonahx replied:
This is a good question, and it's come a few places in your PRs as well as others.  I'll try a short-ish answer here.  Tell me what you think, or if it raises other questions....

> From my point of view, if you have a responsibility that needs to be done in several places then it makes sense to make it a command class.

This is mostly try if:

1. The logic has side-effects.
2. The logic has minimum complexity

The reason I add the 2. caveat is if it's really simple, like 1 line, it might be simpler to repeat it.  Again, introducing an abstraction is not cost-free.  Just like it's possible to have too many tiny methods in a class, it's possible to have too many tiny command classes... you may have removed duplication, but you force the reader to know about more things, and to be able to remember and use them properly.  This is a deep subject, and I think the best way to get a handle on it personally and as a team is to have discussions about multiple specific examples.

> here is the example. we have this class: https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/extract_container_name.rb

I remember seeing this one before. 

So I think my take on this one is that we're missing two things:

1. A value object like `K8sHost`, which is either created with `service_id` and has a `container_name` method, or is created without a `service_id` and has a `container_name(service_id)` method, depending on your needs.
2. A repository object which would return a `K8sHost`, given a `host_id` (or given a `host_id`/`service_id` combo)

We don't use `Repo` objects formally in Conjur, though I have may have created 1 or 2, and we treat the Rails model objects like repos in many places.  But the basic idea is that the repository knows all the database storage details, and it just give you back a nice, clean application object that works the way you want it to.

So in this particular case,  all this junk:

```ruby
 def container_name
        annotation_value("authn-k8s/#{@service_id}/#{AUTHENTICATION_CONTAINER_NAME_ANNOTATION}") ||
          annotation_value("authn-k8s/#{AUTHENTICATION_CONTAINER_NAME_ANNOTATION}") ||
          annotation_value("kubernetes/#{AUTHENTICATION_CONTAINER_NAME_ANNOTATION}") ||
          default_authentication_container_name
      end
```
is just "details about how Conjur happens to store k8s info in its database using annotations".  Your application code does't care about that crap.  That's just database implementation detail.  So we want our Repo, and _only_ our repo, to know about this stuff.  We'd just query it and get back `K8sHost`, say.  

So this stuff fits more naturally at the at lower level than in a `CommandClass` object, which (per this README) are intended for the higher level coordination of multiple tasks into a single use-case.

jonahx commented:
@orenbm and @micahlee I've addressed all the comments except for the naming example (I have some ideas) and integrating the long reply I made to Oren's last comment.  Let me know what you think.
